Avoid duplicate NetworkProcess::AddAllowedFirstPartyForCookies messages

szewai opened the pull request:
#### 0d30af72a1b23a3fbe4c9b2a681b609c75dd7a02
<pre>
Avoid duplicate NetworkProcess::AddAllowedFirstPartyForCookies messages
<a href="https://bugs.webkit.org/show_bug.cgi?id=272731">https://bugs.webkit.org/show_bug.cgi?id=272731</a>
<a href="https://rdar.apple.com/126531726">rdar://126531726</a>

Reviewed by Alex Christensen.

Network process currently receives multiple NetworkProcess::AddAllowedFirstPartyForCookies messages for the same process
identifier and site pair, even though NetworkProcessProxy::addAllowedFirstPartyForCookies asks to not send the message
when the entry already exists. This is caused by some call sites explicitly sending the AddAllowedFirstPartyForCookies
message to network process instead of invoking NetworkProcessProxy::addAllowedFirstPartyForCookies. This patch fixes
that issue.

This patch also fixes a use-after-move issue in WebFrameProxy::prepareForProvisionalNavigationInProcess.

* Source/WebKit/UIProcess/WebFrameProxy.cpp:
(WebKit::WebFrameProxy::prepareForProvisionalNavigationInProcess):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::loadAlternateHTML):
(WebKit::WebPageProxy::continueNavigationInNewProcess):
(WebKit::WebPageProxy::triggerBrowsingContextGroupSwitchForNavigation):
* Source/WebKit/UIProcess/WebProcessPool.cpp:
(WebKit::WebProcessPool::establishRemoteWorkerContextConnectionToNetworkProcess):

Canonical link: <a href="https://commits.webkit.org/277847@main">https://commits.webkit.org/277847@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a79a41d02b6a544f7a6b64e8e5b7129753edde82

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/48703 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/27914 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/51664 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/51390 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/44768 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/33851 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/25444 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/39828 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/49285 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/25571 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/42025 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/20925 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/23046 "Passed tests") | | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/6759 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/44987 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/43691 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/53299 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/23750 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/20081 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/47120 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/25016 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/42230 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/46070 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/10733 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/25820 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/24737 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -7281,11 +7281,12 @@ void WebPageProxy::triggerBrowsingContextGroupSwitchForNavigation(uint64_t navig
-    auto processIdentifier = processForNavigation->coreProcessIdentifier();
-    auto preventProcessShutdownScope = processForNavigation->shutdownPreventingScope();
+    if (!processForNavigation)
+        return completionHandler(false);
```
I'm not sure about this.  On the one hand it makes the code more robust, but on the other hand it takes a condition that would clearly cause a crash and we would look into it and turns it into possibly a mysterious failure that is hard to investigate.  I think I would prefer adding RELEASE_ASSERT(processForNavigation) to keep existing behavior of crashing if it's null.

szewai replied:
`RELEASE_ASSERT(processForNavigation)` will be hit 100% on wpt and gtk bots according to https://github.com/WebKit/WebKit/pull/27310#issuecomment-2059485357; this is how I found `processForNavigation` is null and decided to add the early return.

szewai committed:
Avoid duplicate NetworkProcess::AddAllowedFirstPartyForCookies messages
https://bugs.webkit.org/show_bug.cgi?id=272731
rdar://126531726

Reviewed by Alex Christensen.

Network process currently receives multiple NetworkProcess::AddAllowedFirstPartyForCookies messages for the same process
identifier and site pair, even though NetworkProcessProxy::addAllowedFirstPartyForCookies asks to not send the message
when the entry already exists. This is caused by some call sites explicitly sending the AddAllowedFirstPartyForCookies
message to network process instead of invoking NetworkProcessProxy::addAllowedFirstPartyForCookies. This patch fixes
that issue.

This patch also fixes a use-after-move issue in WebFrameProxy::prepareForProvisionalNavigationInProcess.

* Source/WebKit/UIProcess/WebFrameProxy.cpp:
(WebKit::WebFrameProxy::prepareForProvisionalNavigationInProcess):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::loadAlternateHTML):
(WebKit::WebPageProxy::continueNavigationInNewProcess):
(WebKit::WebPageProxy::triggerBrowsingContextGroupSwitchForNavigation):
* Source/WebKit/UIProcess/WebProcessPool.cpp:
(WebKit::WebProcessPool::establishRemoteWorkerContextConnectionToNetworkProcess):

Canonical link: https://commits.webkit.org/277847@main
